File: Data Structures & Algorithms/trapping-rain-water/submission-0.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let result = 0;

        const n = height.length;

        let maxHeight = 0;
        for (const h of height) {
            maxHeight = Math.max(h, maxHeight);
        }

        for (let i = 0; i < maxHeight; i++) {
            let leftBarIndex = 0;
            while (leftBarIndex !== n) {
                while (leftBarIndex < n) {
                    if (height[leftBarIndex] > 0) {
                        break;
                    }
                    leftBarIndex++;
                }
                if (leftBarIndex === n) {
                    break;
                }
                let rightBarIndex = leftBarIndex + 1;
                while (rightBarIndex < n) {
                    if (height[rightBarIndex] > 0) {
                        break;
                    }
                    rightBarIndex++;
                }
                if (rightBarIndex === n) {
                    break;
                }

                result += rightBarIndex - leftBarIndex - 1;
                leftBarIndex = rightBarIndex;
            }
            for (let j = 0; j < height.length; j++) {
                height[j]--;
            }
        }

        return result;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int l = 0;
        int r = heights.length - 1;
        int maxArea = 0;
        while(l < r) {
            maxArea = Math.max(maxArea, area(heights, l, r));
            if(heights[l] < heights[r]) {
                l++;
            } else {
                r--;
            }
        }

        return maxArea;
    }

    private int area(int[] heights, int l, int r) {
       return (r - l) * Math.min(heights[l], heights[r]);
    }
}
